torch/_decomp/decompositions.py

```diff
@@ -4446,6 +4446,57 @@ def _weight_norm_interface(x, y, dim):
     norm = x.norm(2, keep_dim, keepdim=True)
     return x * (y / norm), norm
 
+@register_decomposition(aten.isin)
+@out_wrapper()
+def isin(elements, test_elements, *, assume_unique=False, invert=False):
+    if test_elements.numel() < 10.0*pow(elements.numel(), 0.145):
+        return isin_default(elements, test_elements, invert=invert)
+    else:
+        return isin_sorting(elements, test_elements, assume_unique=assume_unique, invert=invert)
+
+def isin_default(elements, test_elements, *, invert=False):
+    x = elements.view(*elements.shape, *((1,)*test_elements.ndim))
+    if not invert:
+        cmp = x == test_elements
+    else:
+        cmp = x != test_elements
+    dim = tuple(range(-1, -test_elements.ndim-1, -1))
+    return cmp.any(dim=dim)
+
+def isin_sorting(elements, test_elements, *, assume_unique=False, invert=False):
+    elements_flat = elements.flatten()
+    test_elements_flat = test_elements.flatten()
+    if assume_unique:
+        # This is the same as the aten implementation. For
+        # assume_unique=False, we cannot use unique() here, so we use a
+        # version with searchsorted instead.
+        all_elements = torch.cat([elements_flat, test_elements_flat])
+        sorted_elements, sorted_order = torch.sort(all_elements, stable=True)
+
+        duplicate_mask = torch.empty_like(sorted_elements, dtype=torch.bool)
+        sorted_except_first = sorted_elements[1:]
+        sorted_except_last = sorted_elements[:-1]
+        if invert:
+            duplicate_mask = duplicate_mask.slice_scatter(sorted_except_first !=
+                                                           sorted_except_last,
+                                                           start=0, end=-1)
+        else:
+            duplicate_mask = duplicate_mask.slice_scatter(sorted_except_first ==
+                                                           sorted_except_last,
+                                                           start=0, end=-1)
+        duplicate_mask = duplicate_mask.index_put([torch.tensor(-1)], torch.tensor(invert))
```

Comment:
Inductor really wants to avoid indexed mutations as these can't be fused, so you might get better results with this version

```suggestion
        duplicate_mask = sorted_elements[1:] == sorted_elements[:-1]
        duplicate_mask = torch.constant_pad(duplicate_mask, [0, 1], False)

        if invert:
            duplicate_mask = duplicate_mask.logical_not()
```

but tbh I suspect it's going to be hard to beat the searchsorted version as that version avoids `index_copy`.